File: 31-next-permutation/next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {

        int n = nums.size();

        int i = n - 2;
        // step1 : finding (ith) element which is smaller than (i+1)
        while (i >= 0 && nums[i] >= nums[i + 1]) {
            i--;
        }

        // finding the successor means finding element  from the last which is
        // greater than (ith element)
        if (i >= 0) {
            int j = n - 1;
            while (nums[j] <= nums[i]) {
                j--;
            }
            swap(nums[j], nums[i]);
        }
        // return the suffix part only not the whole array
        reverse(nums.begin() + i + 1, nums.end());
    }
};

File: 31-next-permutation/next-permutation.java
class Solution {
    public void swap(int[] nums, int i, int j) {
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
}

    public void reverse(int[] nums,int l) {
        
        int r = nums.length - 1;

        while (l < r) {
            swap(nums, l,r);
            l++;
            r--;
        }
    }

    public void nextPermutation(int[] nums) {
        int n = nums.length;

        int i = n - 2;

        while (i >= 0 && nums[i] >= nums[i + 1]) {
            i--;
        }

        if (i >= 0) {
            int j = n - 1;
            while (j >= 0 && nums[j] <= nums[i]) {
                j--;
            }
           swap(nums, j, i);
        }

        reverse(nums,i+1);
    }
}